storm-server/src/main/java/org/apache/storm/scheduler/resource/ResourceAwareScheduler.java

```diff
@@ -136,10 +146,24 @@ private void scheduleTopology(TopologyDetails td, Cluster cluster, final User to
             return;
         }
 
+        final IStrategy finalRasStrategy = rasStrategy;
         for (int i = 0; i < maxSchedulingAttempts; i++) {
             SingleTopologyCluster toSchedule = new SingleTopologyCluster(workingState, td.getId());
             try {
-                SchedulingResult result = rasStrategy.schedule(toSchedule, td);
+                SchedulingResult result = null;
+                Future<SchedulingResult> schedulingFuture = backgroundScheduling.submit(
+                    () -> finalRasStrategy.schedule(toSchedule, td)
+                );
+                try {
+                    result = schedulingFuture.get(schedulingTimeoutSeconds, TimeUnit.SECONDS);
+                } catch (TimeoutException te) {
+                    markFailedTopology(topologySubmitter, cluster, td, "Scheduling took too long for "
+                            + td.getId() + " using strategy " + rasStrategy.getClass().getName() + " timeout after "
+                            + schedulingTimeoutSeconds + " seconds.");
+                    schedulingFuture.cancel(true);
+                    rasStrategy.stop();
```

Comment:
Consider dropping the `stop` method and make the scheduler check for interrupts instead. You're already sending an interrupt to it via the `cancel` call above.